[Eugene Ong Wei Xiang] iP

Eugene-Ong-W-X opened the pull request:
# Ren
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes/))

Ren frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is, 
1. download it from [here](https://github.com/Eugene-Ong-W-X/ip/releases/tag/A-Jar).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you :wink:

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

Here's a sneak peek at the `main` method of the next iteration.
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eugene-Ong-W-X committed:
Level-1 Done

Eugene-Ong-W-X committed:
Level-2 Done

Eugene-Ong-W-X committed:
Level-3 Done

Eugene-Ong-W-X committed:
Level-4 Done

Eugene-Ong-W-X committed:
A-TextUiTesting Done

Eugene-Ong-W-X committed:
Level-5 Done

Eugene-Ong-W-X committed:
Level-6 Done

Eugene-Ong-W-X committed:
A-Enums Done

Eugene-Ong-W-X committed:
CheckStyle Done

Eugene-Ong-W-X committed:
Add Level-7 | Read from & Write to File

The program starts with a fresh empty list on startup.

Users will be able to use it as a long-term task manager
if they are able to save their tasks.

Implemented a Storage class that handles reading from and
writing to a File stored in /data/.

Eugene-Ong-W-X committed:
Add Level-8 | Timestamps properly implemented

Time and Date for Tasks are stored as Strings.

This imposes limitations on possible future features like filtering Tasks by date.

TimeStamp class implemented to store date and time information.

Eugene-Ong-W-X committed:
Merge branch 'branch-Level-7'

Eugene-Ong-W-X committed:
Merge branch 'branch-Level-8' into HEAD

Merge Conflict resolved in Duke.java

Eugene-Ong-W-X committed:
More OOP | Add Ui, Parser, TaskList classes

Alot of code resides in Duke.java

Related functions should be bundled into new classes so
future changes will be easier to implement.

Extracted Ui, Parser, TaskList classes from Duke.java
Now, Duke.java does not hold any complicated code.

Eugene-Ong-W-X committed:
Rename Duke/Alfred to Ren and Add 2 packages

Name alternated between Duke and Alfred.

Confusing for users.

Name has been standardized to Ren.
ren package and ren.task package added.

Eugene-Ong-W-X committed:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2223S1/ip

Eugene-Ong-W-X committed:
Add Gradle Support

Eugene-Ong-W-X committed:
Add JUnit Tests

Added JUnit Tests for 6 out of 11 classes.
Classes without JUnit Tests:
* Ren
* Ui
* TaskList
* Storage
* Tasks

Eugene-Ong-W-X committed:
Release v0.2 (Packaging as JAR file)

Eugene-Ong-W-X committed:
Add JavaDoc

Add JavaDoc for all non-private classes/methods
Add JavaDoc for all non-trivial private methods.

Eugene-Ong-W-X committed:
Java Standards: Fixed test codes

Fixed test codes to comply with Java Standards.

Eugene-Ong-W-X committed:
Add Find Feature

Users are unable to easily find a task from a dense list of tasks.

Users are discouraged from using Ren to track many tasks.

Users can now search for tasks with keywords.

Eugene-Ong-W-X committed:
Merge branch 'branch-A-JavaDoc'

Eugene-Ong-W-X committed:
Merge branch 'branch-A-CodingStandard'

Eugene-Ong-W-X committed:
Merge branch 'branch-Level-9'

Eugene-Ong-W-X committed:
Use CheckStyle

1 error caught in ParserTest.

Eugene-Ong-W-X committed:
Merge branch 'branch-A-CheckStyle'

cowlinn commented on src/main/java/DukeException.java:
```diff
@@ -0,0 +1,10 @@
+
+    @Override
+    public String toString() {
+        return " ☹ Apologies! "+ super.getMessage() + "\n";
```
In general, I really liked how you separated your toStrings via line breaks to improve readability! 

Small comment here in line 8 that there should be a spacing after the string, so
`" ☹ Apologies! " +`

cowlinn commented on src/main/java/Duke.java:
```diff
@@ -48,44 +52,50 @@ private static void addTask(TaskType type, String task, String dateTime) {
+    private void updateTask(TaskAction func, int taskNum) throws DukeException {
         if (taskNum <= pointer && taskNum > 0) {
             switch (func) {
             case MARK:
```
Really good use of enums! Properly indented as well might just steal it heh

cowlinn commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,181 @@
+        for (int i = 0; i < this.counter; i++) {
+            String[] data = dataArray.get(i).split("\\|");
+            Task newTask = null;
+            switch (data[0]) {
```
Perhaps data[0] could be given a more meaningful name, perhaps something like: 
```
String dataHeader = data[0];
switch (dataHeader) {
```
My loading is very similar to yours and I just realized it might be a bit confusing for people who didn't split their save data by lines. :)

cowlinn commented on src/main/java/ren/TimeStamp.java:
```diff
@@ -22,6 +22,7 @@ private TimeStamp(LocalDateTime timestamp) {
      * @throws RenException If dateTime is not formatted correctly.
      */
     public static TimeStamp of(String dateTime) throws RenException {
```
This is completely up to you but I felt the method name for this factory method is a little short? Perhaps something like 
`public static TimeStamp **generateTimestamp**(String dateTime) throws RenException` could be something a bit more descriptive? I do understand that "of" is used for the grammar to flow in another context, so feel free to ignore this!

cowlinn commented on src/main/java/Deadline.java:
```diff
@@ -1,6 +1,15 @@
+/**
```
Sorry I just saw you had Checkstyle added so ignore all the previous mentions of spacing issues, they should be resolved by here. Regardless formatting wise I mean there aren't any styling violations apart from some small naming potential improvements! Really good job man :)

cowlinn reviewed (commented):
Heyo not sure if I'm doing this correctly but lemme know if you have any questions! Not much problems with formatting that I can find haha

Eugene-Ong-W-X committed:
Upgrade to GUI from Text UI

Ren runs on a text ui.

This is unlikely to be pleasing to users.

Let's upgrade the ui to a GUI that resembles a chat app.

This would make the program more pleasing to use.

Eugene-Ong-W-X committed:
Merge branch 'branch-Level-10'

Eugene-Ong-W-X committed:
Add and enable Assertions

Currently, important assumptions are checked using if statements
or not checked at all.

This is dangerous as it is not guaranteed that the assumptions will
hold. Additionally, there is a risk that some assumptions might be
forgotten about.

Now, all important assumptions are checked using Assertions.

Assertions allow assumptions to be clearly marked out from other code.

Eugene-Ong-W-X committed:
Improve Code Quality

Currently, the code does not strictly follow the newly introduced
Code Quality guidelines.

This indicates areas of improvements in the code readability.

Now, after combing through the code with the new guidelines,
several changes have been made to improve code readability.

Eugene-Ong-W-X committed:
Merge pull request #2 from Eugene-Ong-W-X/branch-A-CodeQuality

Improve Code Quality

Eugene-Ong-W-X committed:
Merge branch 'master' into branch-A-Assertions

Eugene-Ong-W-X committed:
Merge pull request #1 from Eugene-Ong-W-X/branch-A-Assertions

Add and enable Assertions

Eugene-Ong-W-X committed:
Set up Java CI using Github Actions

Eugene-Ong-W-X committed:
Fix for Java CI implementation

Eugene-Ong-W-X committed:
Fix 2 for Java CI implementation

Eugene-Ong-W-X committed:
Add Sort Feature

Currently, the list of tasks is not sorted and left in the order the
user added them.

This hampers the user's ability to organise their tasks by certain
characteristics.

Now that the sort feature is implemented, the user can sort tasks by:
* their type (Todo > Deadline > Event)
* their status (Unmarked Tasks > Marked Tasks)
* their description (sorted lexicographically while ignoring case)
* their date (sorted chronologically while Todo's are sorted last)

Eugene-Ong-W-X committed:
Merge branch 'branch-BCD-Extension'

Eugene-Ong-W-X committed:
Set up Ren Website

Product Screenshot included.
User Guide v0.1 included.

Eugene-Ong-W-X committed:
Update README.md

Eugene-Ong-W-X committed:
Fix Code Quality
